observability/src/main/java/io/grpc/observability/logging/GcpLogSink.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2022 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.observability.logging;
+
+import com.google.cloud.MonitoredResource;
+import com.google.cloud.logging.LogEntry;
+import com.google.cloud.logging.Logging;
+import com.google.cloud.logging.LoggingOptions;
+import com.google.cloud.logging.Payload.JsonPayload;
+import com.google.cloud.logging.Severity;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.protobuf.util.JsonFormat;
+import io.grpc.internal.JsonParser;
+import io.grpc.observabilitylog.v1.GrpcLogRecord;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Sink for Google Cloud Logging.
+ */
+public class GcpLogSink implements Sink {
+  private final Logger logger = Logger.getLogger(GcpLogSink.class.getName());
+
+  private static final String SERVICE_TO_EXCLUDE = "google.logging.v2.LoggingServiceV2";
+  private static final String DEFAULT_LOG_NAME = "grpc";
+  private static volatile GcpLogSink gcpLogSinkInstance = null;
+  private final Logging gcpLoggingClient;
+
+  /**
+   * Constructor for custom sink.
+   */
+  private GcpLogSink(String projectId) {
+    LoggingOptions.Builder builder = LoggingOptions.newBuilder();
+    if (!Strings.isNullOrEmpty(projectId)) {
+      builder.setProjectId(projectId);
+    }
+    gcpLoggingClient =  builder.build().getService();
+  }
+
+  private GcpLogSink(Logging client) {
+    gcpLoggingClient = client;
+  }
+
+  @VisibleForTesting
+  public static void setInstance(Logging loggingClient) {
+    gcpLogSinkInstance = new GcpLogSink(loggingClient);
+  }
+
+  /**
+   * Retrieves a single instance of GcpLogSink.
+   *
+   * @param destinationProjectId cloud project id to write logs
+   */
+  public static GcpLogSink getInstance(String destinationProjectId) {
+    if (gcpLogSinkInstance == null) {
+      synchronized (GcpLogSink.class) {
+        if (gcpLogSinkInstance == null) {
+          gcpLogSinkInstance = new GcpLogSink(destinationProjectId);
+        }
+      }
+    }
+    return gcpLogSinkInstance;
+  }
+
+  @Override
+  public synchronized void write(GrpcLogRecord logProto) {
```

Comment:
why does the whole method need to be synchronized? Only `gcpLoggingClient.write` needs to be serialized as far as I can see

Reply:
Done